docs/universal-gateway/examples/front-door-pattern.mdx

```diff
@@ -0,0 +1,94 @@
+---
+title: "Implement the 'Front Door' Pattern"
+description: By using Cloud Endpoint and internal Agent Endpoints together, you loosely double your gateway and services for extra flexibility.
+sidebar_label: "The Front Door Pattern"
+---
+
+import ReserveDomain from "./snippets/_reserve-domain.mdx";
+import CloudEndpoint from "./snippets/_cloud-endpoint.mdx";
+import Back from "./snippets/_back-to-examples.mdx";
+
+The "front door" pattern involves using a Cloud Endpoint, the forward-internal Traffic Policy action, and an internal Agent Endpoint to route traffic from the public internet to your service(s).
+
+With this pattern, you can:
+
+- Host any number of services under a single hostname (e.g. `your-company.com`).
+- Decouple gateway configuration from per-service traffic management policies.
+- Prevent any driect exposure of your services or their host systems to the public internet.
+- Route traffic and add authentication to microservices with a single API gateway configuration.
+- Authenticate developer environments or CI preview builds behind a single endpoint.
+
+## 1. Create an endpoint for your service
+
+Start an internal [Agent Endpoint](/docs/universal-gateway/agent-endpoints/), replacing `$PORT` based on where your service listens.
+You can also use one of our [SDKs](/docs/agent/sdks) or the [Kubernetes Operator](/docs/k8s).
+
+```bash
+ngrok http $PORT --url https://service.internal
+```
+
+## 2. Reserve a domain
+
+<ReserveDomain />
+
+## 3. Create a Cloud Endpoint
+
+<CloudEndpoint />
+
+## 4. Add routing to your service with Traffic Policy
+
+Copy and paste the policy below into your new Cloud Endpoint in the ngrok dashboard.
+
+```yaml
+on_http_request:
+  - actions:
+      - type: forward-internal
+        config:
+          url: http://service.internal
+```
+
+### What's happening here?
+
+On every HTTP request, the Traffic Policy engine forwards the traffic to the
+internal agent endpoint you created at `https://service.internal`.
+
+## 5. Add a second service and routing (optional)
+
+If you have another service to host under the front door pattern, start another agent.
+
+```bash
+ngrok http $PORT --url https://service-two.internal
+```
+
+## 5a. Update your routing based on paths
+
+Copy and paste the policy below into the ngrok dashboard, replacing `/one` and `/two` with the paths you'd like to use for routing traffic to each service.
+
+```yaml
+on_http_request:
+  - expressions:
+      - "req.url.includes('/one')"
+    actions:
+      - type: forward-internal
+        config:
+          url: http://service.internal
+  - expressions:
+      - "req.url.includes('/two')"
+    actions:
+      - type: forward-internal
+        config:
+          url: http://service-two.internal
+```
+
+## 6. Try out your endpoint!
+
+- In your browser: https://your-service.ngrok.app
+- In your terminal: `curl https://your-service.ngrok.app`
```

Comment:
```suggestion
Visit your reserved domain either in the browser or in the terminal using a tool like `curl`. You should see the app or service at the port connected to your internal endpoint. 
```
Feel free to modify this suggestion. The spirit of it is that we don't know what their URL will be, and people skimming this might just copy this placeholder URL you've written, paste it in the browser, see nothing working, and get a bit confused for a second. I think it would be better to just explicitly say that they should use their reserved domain